### ISSUE TITLE
[Question] Multi-head attention init in Tutorial 6

### ISSUE THREAD

**Issue by pi-tau** [phlippe/uvadlc_notebooks#92]
**Tutorial**: 6

**Describe the bug**
Wouldn't say it is a bug, but something that I find strange and wanted to bring up.
For the multi-head attention, I really like the explanation you give about scaling the attention logits before applying the softmax in order to keep the variance manageable. From what I understand we want to keep the variance of keys, queries (and values?) close to 1 and that is why you would initialize the $W_Q$, $W_K$ and $W_V$ matrices using `Xavier` initialization.

In Cell 5 we use `Xavier` initialization but we initialize the entire `qkv_proj` which holds all three matrices for all heads.
I think that it would be more in line with theory if we initialize this way:

```python3
nn.init.normal_(self.qkv_proj, mean=0., std=np.sqrt(2 / (input_dim + embed_dim // n_heads)))
```

With large `embed_dim` and small `n_heads` I don't think it really makes that much difference, but I would be happy to hear your thoughts about it.

**Additional context**
As always great content. Thanks a lot for sharing : )

**Comment by phlippe** [comment]
Hi, thanks for your interesting question! You are right that if we want to keep the variance at exactly 1, the weight matrices need to be initialized slightly different. The scaling of the attention logits is mostly just needed to prevent the variance from becoming too large, and the Layer Norm and Residual connections take care of caring the gradients and variance through the network (i.e. we don't have to worry about 'losing' variance). I mostly decided to use Xavier uniform here since this is what most official implementations use (e.g. native flax), but I agree that the stacking of the weight matrices slightly changes it. Since models are usually not too sensitive to this initialization, I added a comment before the layer in 01b1d43f8df09483c3ee9ee5af0e377e52070657 to make this clear. :)
